File: packages/template-ts/src/value/stringify.ts
// VAL-8: conversion of a value to text.
import { numberToString } from './number.js';
import { SafeString, type Value } from './value.js';

export class StringifyError extends Error {
  constructor() {
    super('a list or map cannot be converted to text');
    this.name = 'StringifyError';
  }
}

// Returns the text of a value, or throws StringifyError for a list or map.
export function stringify(value: Value): string {
  if (value === null) return '';
  if (typeof value === 'boolean') return value ? 'true' : 'false';
  if (typeof value === 'number') return numberToString(value);
  if (typeof value === 'string') return value;
  if (value instanceof SafeString) return value.text;
  throw new StringifyError();
}

File: packages/template-go/value/stringify.go
package value

import "errors"

// ErrStringify is returned when a list or map is converted to text (VAL-8).
var ErrStringify = errors.New("a list or map cannot be converted to text")

// Stringify implements VAL-8.
func Stringify(v Value) (string, error) {
	switch x := v.(type) {
	case nil:
		return "", nil
	case bool:
		if x {
			return "true", nil
		}
		return "false", nil
	case float64:
		return NumberToString(x), nil
	case string:
		return x, nil
	case SafeString:
		return x.Text, nil
	}
	return "", ErrStringify
}
